src/llmq/quorums_dkgsessionhandler.cpp

```diff
@@ -138,6 +138,10 @@ void CDKGSessionHandler::ProcessMessage(CNode* pfrom, const std::string& strComm
 
 void CDKGSessionHandler::StartThread()
 {
+    if (phaseHandlerThread.joinable()) {
+        throw std::runtime_error("Tried to start an already started CDKGSessionHandler thread.");
+    }
+
     auto threadName = [&]() -> auto {
```

Comment:
This whole function object can be replaced by a simple `auto threadName = strprintf("q-phase-%d", params.type);`
There is no need to use a function object here, as the result is already known/calculatable before the thread starts and never changes.

Reply:
@codablock it even was like this before but by adding the thread to the `TraceThread` it started to show a very weird string related behaviour where i couldn't find an answer in reasonable time so i decided to do it the way it is right now as workaround as it really started to annoy me lol. Since there was the side effect of having a check for valid LLMQ types in the `switch` i was kind of happy with the solution :D 

Anyway, im open to apply your suggestion if you can provide me a fix for the following: 

```
    auto threadName = strprintf("q-phase-%d", params.type);

    phaseHandlerThread = std::thread(&TraceThread<std::function<void()> >, threadName.c_str(), std::function<void()>(std::bind(&CDKGSessionHandler::PhaseHandlerThread, this)));
```

The above prints: 
<img width="432" alt="Bildschirmfoto 2020-07-14 um 11 41 15" src="https://user-images.githubusercontent.com/35775977/87412377-45e5ae00-c5c9-11ea-9670-8e3bf89cad3f.png">
 
Where my approach prints: 
<img width="410" alt="Bildschirmfoto 2020-07-14 um 12 01 23" src="https://user-images.githubusercontent.com/35775977/87412862-d2906c00-c5c9-11ea-8736-980f5ee552fb.png">

If you or someone else has an idea/fix let me know!

Reply:
I guess this to happen because the `c_str()` ptr is forwarded to the thread and the string object of `threadName` doesn't exist when this gets used and in my approach it returns just a char array?

Reply:
Yeah you're right. The issue got fixed due to the use of plain C strings and not by the use of function objects.

The proper fix would be to change the method `TraceThread` to take a std::string (not a reference!!) object instead of a char* pointer.

Reply:
Right that would probably work.. for some reason i didn't even think about changing the `TraceThread` lol.. anyway, do we need this fix here now as its kind of unrelated or are we good with the function approach? I mean, isn't it nice to also be sure we don't start a thread for an arbitrary value different out of the range of valid LLMQ types? :)

Reply:
The `sprintf` is just better haha. Pushed a fix in 0f20246 but this requires #3609 to make it working as expected.